### PR TITLE
[codex] add fused schedule and bound compiled GPU memory

### PR DESCRIPTION
**Pull request by ooples**
## Summary
- Add a public `WarmupDecayMode` plus `LrSchedule.LinearWarmup(...)` for compiled/fused optimizer schedules.
- Match the AiDotNet `LinearWarmupScheduler` batch sequence for constant, linear-decay, and cosine-decay post-warmup behavior.
- Reject invalid schedules where total steps are less than warmup steps.

## Verification
- `dotnet test tests\AiDotNet.Tensors.Tests\AiDotNet.Tensors.Tests.csproj --framework net10.0 --filter FullyQualifiedName~AiDotNet.Tensors.Tests.Engines.Compilation.LrScheduleTests` passed 24/24.
- Consumed via local package in HarmonicEngine with AiDotNet mapping changes: fused package canary passed and showed training movement (`avg_loss 9.4550 -> 7.4261`, `pL1 40621.47 -> 46411.97`, no eager fallback).

## Summary by CodeRabbit

* **New Features**
  * Added `LrSchedule.LinearWarmup(...)` with `WarmupDecayMode` (Constant, Linear, Cosine).
  * Added `ProximalL1Update` optimizer support across GPU backends (proximal-gradient / ISTA with L1 soft-thresholding).

* **Bug Fixes**
  * Improved activation-cache cleanup and eviction behavior during compiled training steps.
  * Fixed GPU allocation tracking to avoid overcounting duplicate allocations.
  * Corrected embedding lookup index handling (int indices) and improved related kernel execution.
  * Hardened BLAS packed-operation routing and attention-gradient buffer initialization.

* **Tests**
  * Expanded learning-rate, GPU residency, and cleanup regression coverage; added optional memory repro test.

### CODE REVIEW

**Comment by ooples** [comment]
Update after the HRE Cortex audit:

- Added a manual opt-in CUDA repro for the Cortex-like `B=1024/S=64/D=256/FF=512/L=4` compiled training memory envelope.
- Generalized compiled eager-step activation cleanup to the FP32 path and protects returned gradients/loss state while dropping dead transients.
- Fixed `GpuMemoryTracker.OnAlloc` live-byte accounting when a reused pointer replaces an existing live record.

Validation:
- Default-skip targeted test passes.
- Opt-in repro with `AIDOTNET_CUDA_GRAPH_STEP=0`, `AIDOTNET_GPU_MEMTRACK=1`, `AIDOTNET_REPRO_STEPS=4`, `AIDOTNET_REPRO_MAX_LIVE_MB=1024`, `AIDOTNET_REPRO_MAX_PEAK_MB=3500` passes with loss decreasing and live memory returning near 516 MiB.
- Packed local `AiDotNet.Tensors.0.104.9-capture.nupkg` and used it in the HRE proof.

**Comment by ooples** [comment]
Codex update 2026-07-01:

Added commit `f1d9db9` for the GPU optimizer residency bug found from the HarmonicEngine Cortex run.

Root cause: generic GPU optimizer paths could update resident GPU optimizer tensors without advancing the tensor `Version` / `_gpuBufferVersion` invariant, which left downstream code with stale/current ambiguity during longer training and post-training evaluation.

Fix:
- Mark GPU-updated optimizer tensors current after dense and sparse optimizer steps.
- Remove deferred materializer entries when activation-cache keys are explicitly freed.
- Keep the Cortex transformer repro test net471-compatible.

Validation:
- `dotnet build src\AiDotNet.Tensors\AiDotNet.Tensors.csproj --no-restore ...` passed.
- net10 `GpuPinnedLifetimeTests|OptimizerKernelsTests`: 26/26 passed.
- net10 leak/compiled-backward/tier3 filter: 21 passed, 2 skipped.
- net471 `GpuPinnedLifetimeTests|OptimizerKernelsTests`: 26/26 passed.
- Packed local `AiDotNet.Tensors.0.104.16-capture.nupkg` and reran the HRE proof against it.

**Comment by ooples** [comment]
Rebase validation complete: local focused tests and hosted build checks are green.

**Comment by ooples** [comment]
﻿Rebased validation details for head 79b584b:

- Package build: dotnet build src\AiDotNet.Tensors\AiDotNet.Tensors.csproj -c Release passed after the license-validator helper was restored/built.
- BLAS hosted-failure repro: net10.0 NAxisParityTests + MachineKernelGemmTests + MachineCodeKernelTests with XPlat Code Coverage passed 78/78.
- Optimizer residency tests: net10.0 GpuPinnedLifetimeTests + OptimizerKernelsTests passed 26/26.
- Optimizer residency tests: net471 GpuPinnedLifetimeTests + OptimizerKernelsTests passed 26/26.
- Leak/backward memory filter: net10.0 GradientTapeLeakTests + CompiledBackwardWalkTests + Tier3MemoryLeaksTests passed 29/29 with 1 expected diagnostic skip.

Hosted checks are green: build, avx512-verify, gemm-bench, and CodeRabbit. The PR is mergeable again.

File: src/AiDotNet.Tensors/Engines/Compilation/LrSchedule.cs [hunk 1]
```diff
@@ -285,3 +318,62 @@ public override double GetLr(int step)
...
+        double warmupInitLr, WarmupDecayMode decayMode, double endLr)
+    {
+        if (warmupSteps < 0) throw new ArgumentOutOfRangeException(nameof(warmupSteps));
+        if (decayMode != WarmupDecayMode.Constant && totalSteps < warmupSteps)
+            throw new ArgumentOutOfRangeException(nameof(totalSteps),
+                "totalSteps must be >= warmupSteps for decay modes.");
+        _lrMax = lrMax;
+        _warmupSteps = warmupSteps;
+        // Eager ctor: _totalSteps = totalSteps > 0 ? totalSteps : warmupSteps.
+        _totalSteps = totalSteps > 0 ? totalSteps : warmupSteps;
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/Compilation/LrSchedule.cs]: Verified addressed in current HEAD (085a311d): LinearWarmupLr's ctor now normalizes first (`int normalizedTotalSteps = totalSteps > 0 ? totalSteps : warmupSteps`) and only then validates `normalizedTotalSteps < warmupSteps` for non-Constant decay, so totalSteps<=0 with warmupSteps>0 no longer throws for Linear/Cosine — it maps to warmupSteps as the eager scheduler documents. Resolving.

[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/Compilation/LrSchedule.cs]: Verified addressed in current HEAD: totalSteps is normalized (totalSteps<=0 => warmupSteps) before the decay-mode validation, matching the eager scheduler. Duplicate of the other LrSchedule totalSteps thread. Resolving.

File: src/AiDotNet.Tensors/Engines/Compilation/LrSchedule.cs [hunk 2]
```diff
@@ -285,3 +318,62 @@ public override double GetLr(int step)
...
+    public override double GetLr(int step)
+    {
+        // The eager LinearWarmupScheduler reports its ctor value (warmupInitLr) on
+        // batch 1 (unfloored), then on batch n>=2 reports max(endLr, ComputeLearningRate(n-1))
+        // — _currentStep is (batch - 1) and Step() floors with _minLearningRate = endLr.
+        // The 1-indexed fused step maps batch n → GetLr(n); reproduce both branches so the
+        // per-step LR is bit-identical to the eager replay (FusedLrScheduleMappingTests).
+        if (step <= 1) return _warmupInitLr;
+        double raw = ComputeRaw(step - 1);
+        return raw > _endLr ? raw : _endLr;
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/Compilation/LrSchedule.cs:360]: Verified addressed in current HEAD: LinearWarmupLr.GetLr returns `_warmupSteps > 0 ? _warmupInitLr : _lrMax` at step 1, so the no-warmup (warmupSteps==0) case starts at lrMax — consistent with the LinearWarmupCosine convention. Resolving.

File: src/AiDotNet.Tensors/Engines/Gpu/GpuOptimizer.cs [hunk 3]
```diff
@@ -33,6 +33,34 @@ namespace AiDotNet.Tensors.Engines.Gpu;
 /// </remarks>
 public static class GpuOptimizer
 {
+    private static void MarkGpuUpdated(Tensor<float>? tensor)
+    {
+        if (tensor is null) return;
+        tensor.MarkModified(syncPoint: null);
+        tensor._gpuBufferVersion = tensor.Version;
+    }
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/Gpu/GpuOptimizer.cs]: Verified addressed in current HEAD: GpuOptimizer now calls `tensor.MarkModified(CreateWriteSyncPoint(backend))` rather than MarkModified(null), so the write fence is recorded and Tensor.Synchronize()/LastWriteSync correctly wait for the in-place optimizer update on async backends. Resolving.

File: src/AiDotNet.Tensors/Engines/DirectGpuTensorEngine.cs [hunk 4]
```diff
@@ -2022,6 +2022,11 @@ internal void FreeFloatActivation(Tensor<float> t)
 
     private bool TryFreeActivationByKey(object key)
     {
+        // Dead tensor: discard the lazy GPU->CPU materializer before freeing
+        // the cache-owned buffer. Materializing here would be wasted work, and
+        // leaving the registration behind would point a later bulk drain at a
+        // freed buffer.
+        Helpers.DeferredArrayMaterializer.Remove(key);
         if (!_activationCache.TryRemove(key, out var e)) return false;
         System.Threading.Interlocked.Add(ref _currentActivationCacheBytes, -e.Buffer.SizeInBytes);
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/DirectGpuTensorEngine.cs:2031]: Verified addressed in current HEAD: the free/invalidate path was centralized into InvalidateActivationCacheEntry (single buffer-disposal chokepoint). A pending deferred materializer is force-materialized to the host BEFORE the buffer is freed, and DeferredArrayMaterializer.Remove is only reached on a materialize failure — a still-valid materializer registration is no longer dropped when the cache entry isn't removed. Resolving.

File: src/AiDotNet.Tensors/Engines/Compilation/CompiledTrainingPlan.cs [hunk 5]
```diff
@@ -1443,20 +1438,22 @@ private Tensor<T> StepEager()
...
+                {
+                    MaterializeStepLoss();
+                }
+                finally
+                {
+                    stepActivationEngine.EvictActivationsCreatedAfter(
+                        stepActivationSnapshot,
+                        BuildStepEvictionProtectSet(),
+                        materializePending: true);
+                }
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/Compilation/CompiledTrainingPlan.cs:1456]: Verified addressed in current HEAD: the StepEager-path EvictActivationsCreatedAfter call now passes `materializePending: false`, so evicting activation-cache entries no longer eagerly triggers every pending GPU->CPU download. Resolving.

File: tests/AiDotNet.Tensors.Tests/Engines/DirectGpu/GpuOptimizerResidencyMockTests.cs [hunk 6]
```diff
@@ -0,0 +1,299 @@
+// Copyright (c) AiDotNet. All rights reserved.
+
+#if !NETFRAMEWORK
+#nullable disable
+
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tensors.Tests/Engines/DirectGpu/GpuOptimizerResidencyMockTests.cs]: Verified addressed in current HEAD: the residency/version regression suite is no longer compiled out on net471 (no #if guard or csproj exclusion remains), so the coverage runs across the supported TFMs. Resolving.

File: src/AiDotNet.Tensors/Engines/DirectGpu/IDirectGpuBackend.cs [hunk 7]
```diff
@@ -2314,6 +2314,17 @@ void NadamUpdate(IGpuBuffer param, IGpuBuffer gradient, IGpuBuffer m, IGpuBuffer
...
+    /// <summary>
+    /// Proximal gradient (ISTA) step with L1 soft-thresholding.
+    /// </summary>
+    /// <param name="param">Parameters to update (in-place).</param>
+    /// <param name="gradient">Gradient values.</param>
+    /// <param name="learningRate">Learning rate.</param>
+    /// <param name="l1Strength">L1 shrinkage threshold.</param>
+    /// <param name="size">Number of parameters.</param>
+    void ProximalL1Update(IGpuBuffer param, IGpuBuffer gradient,
+        float learningRate, float l1Strength, int size);
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/DirectGpu/IDirectGpuBackend.cs:2326]: Verified addressed in current HEAD: CUDA is in the ProximalL1Update parity set — CudaBackend implements both ProximalL1Update and SparseProximalL1Update (CudaBackend.cs), with the proximal_l1_update kernel in Kernels/CudaOptimizerKernels.cs, alongside HIP/OpenCL/Metal/Vulkan/WebGPU. Resolving.
